### PR TITLE
[WIP] Introduce Repository::Model

### PR DESCRIPTION
**Pull request by solnic**
This adds a new type of repo that is designed to provide a simple CRUD interface for a single "model" which maps directory to the canonical relation schema (ie a table in an sql db).

API:

``` ruby
class UserRepo < ROM::Repository::Model[:users]
end

user_repo = UserRepo.new(rom)

user_repo.create(name: 'Jane')

user_repo.update(1, name: 'Jane Doe')

user_repo.delete(1)
```

refs #13

### CODE REVIEW

**Comment by robinetmiller** [comment]
Problem 1: Something that occurred to me since #13 is what about when they need to expand to multi-relation situation? Repos handle one _or more_ relations at a time. 

Relatedly, what about association methods? eg. Where does a create user_and_tasks live? In UserRepository? TaskRepository?

Problem 2: natural keys. Using numbered indexes is common, but it strikes me that may only be because of ORM limitations forcing a standardization. If one of ROM's strengths is purportedly flexibility, it should be able to handle custom keys. 

Problem 3: Introduces a new part. The ideal library has a minimal set of interaction points, and the proposal above is adding another class. I can already hear people asking "What's the difference between ROM::Repository and ROM::Repository::Model?". I don't know myself. Would we use both? 

What about something like:

``` ruby
class MyRepo < ROM::Repository
   relations :users, :tasks

end

repo = MyRepo.new(rom)

# repo uses the first param to know which relation to use
# from the developer perspective, they're just specifying what type of domain thing to persist. 
repo.create(:users, name: 'Jane')
repo.create(:tasks, name: 'Laundry')

repo.update(:users, 1, name: 'Jane Doe')
repo.update(:tasks, 1, name: 'Dry Cleaning')

repo.delete(:tasks, 1)
```

This doesn't change what the developer needs to know about, and it provides a natural place for associations custom methods to live: in the repo. It also doesn't require any changes to existing code (not that that's a huge deal in a young library like ROM).

**Comment by solnic** [comment]
My idea was to make it clear that it's a repo for a single relation. I don't have a strong opinion though. What you suggest will result in more complicated objects but it wouldn't be a big deal from the usage pov. Let's see what others have to say. This feature is mostly for easy on-boarding so that you can quickly start using rom and to cover basic use cases where simple crud is all you need. As far as associations go - you are on your own, composing a command graph for nested input is the simplest solution but with a repo you could process input manually. We could think about some automation for command graphs in the near future too.

**Comment by kwando** [comment]
The idea of having simple CUD operations available on a ROM::Repository is great. Having those CUD operations available only on repos with a single relation should be fine I guess.

I just had a crazy idea, which I do not have been thinking through completely yet so it might not make sense.

This is regarding point 1 Robin raised above.

What if we could have "virtual relations" / "views" somehow in rom, then we could build up a "user_with_tasks"-relation, and register the needed CUD command graphs for that virtual relation?

**Comment by gotar** [comment]
I like idea with having repo per relation, like in @solnic example. Adding sth more complex will always be possible (manually). For default looks good.

**Comment by solnic** [comment]
@kwando yeah this is an idea that goes back to DM2. You can already built up a relation graph, so it's a matter of reflecting on graph and building a schema from it in order to know what kind of commands are needed. It's not only possible, but also something I wanted to explore eventually. This is probably something that can be done later this year.

**Comment by solnic** [comment]
Thanks guys for the feedback. I'm gonna see how it could be done like @robinetmiller suggests and push it in this PR for comparison (while keeping `Repository::Model`)

**Comment by endash** [comment]
Bear with me as this is kinda philosophical at points. Sorry for the length. Also this may or may not seem apropos of nothing, but I'm incorporating my thoughts on the broader notion of repositories and ROM that have percolated through previous discussions.

My main problem in these discussions is that I'm never quite sure what different people mean by the word "repository." There seem to be two different takes: (1) a low-level interface more or less isomorphic to the database, in terms of exposing CRUD actions and composing relations for more complex capabilities. (2) a high-level, domain-centered interface that hides those things from the app entirely while still clearly fulfilling the role of persistence mechanism. How appropriate this feature seems, and what relation it has to more complex apps/needs, will vary partly based on which camp you're in.

Complexity is the right lens through which to view things, but it isn't  necessarily true that simple use cases are just degenerate versions of more complex use cases, along a smooth continuum. By analogy to physics, for instance, the things we take for granted about the world (like separate electromagnetic and weak forces) are actually only true at low energies. At high energies the distinctions we make and that are so important to our world begin to disappear—but that doesn't make electromagnetism a degenerate version of the weak force. I believe these two notions of a "repository" are related in a similar way, except a bit opposite: for very simple use cases they merge into one, and only at higher levels of complexity do they separate into very different beasts. It's what happens when you extrapolate one view appropriate at low complexity to higher complexities where things can go wonky.

The first kind of repository might involve a method call to create a user, something like `user_repo.create(name: "Alice", admin: true, project_roles: [...])`. The second kind, more domain-oriented and less concerned specifically with interacting with a database, might involve a method call more like `user_repo.create_admin_for_projects("Alice", project1, project2)`. In the first case, our HABTM relationship between `User` and `Project` is a join table, and as such reflected as a relation itself, and finally it bubbles up into an app-level concern when you need to create a user. In the second case, we provide a highly specialized method with data and expect it to know the implementation details of creating the join records.

For our purposes complexity can mostly, but not entirely, be defined in terms of composing relations. When dealing purely with single relations, the two versions of "repository" can be more or less treated identically—all you can really do with only a single relation is create a new record, update an existing record, or delete a record. The fact that these are isomorphic to the database's CRUD capabilities is coincidence. One could engineer scenarios where single-relation entities were appropriately represented via higher level actions (`create_admin` rather than just `create(admin: true)`), but more often than not that's more symptomatic of an overly conflated domain model than actual required complexity in the presistence interface. A high level and a low level interface just happen to look identical, at low complexities.

If you extrapolate from that superficial equivalence to a repository that tries to tie together 2 or 3 relations, knowing their associations and schemas, and inferring what sort of actions you might want in various combinations then it'll quickly start to look like an unmanageable, short-lived convenience that doesn't scale when you need it to, or remains very low-level when you expect something higher-level. This can seem very frustrating and pointless. If, on the other hand, you take the second view, you exploit the simplicity for as long as you can, accepting that it won't—CAN'T—scale—that no library is going to be able to compose your relations into high level semantics, translating trivial issues in the app into database detail, and vice-versa. It becomes a simple matter of X being appropriate for these use cases and Y being appropriate for other use cases.

So the questions become, in my mind: Are there enough use cases at that low level of complexity to justify a quick-and-easy one-size-fits-all CRUD repo? Without question: absolutely. Does the CRUD repo scale? No. Should it? No. Are some people going to hit that barrier more quickly than anothers? Yes. Is it still worth having it, despite its explicit lack of universality? Absolutely. Are there other uses for low-level functionality that can help with complex compositions? Yes, but not as part of the repository interface.

My issue with @robinetmiller's suggestion is that it explictly endorses the view of a repository interface as dealing in low-level semantics. If you're dealing with a single user model, you should have a user repository. `#create` by itself, as a generic method we can handle automatically, doesn't make any sense when dealing with more than one relation. If you have to tell it what to create you just have a thin facade in front of your database, not a repository. 

It's like that old quasijoke about making something as simple as possible, but no simpler—the class as it stands is already as low-level as it can be. Any lower, and it destroys the high level semantics which are what we're actually trying to provide a convenience for.

tl;dr — a simple CRUD interface for single relations makes sense because while the semantics are superificially low-level, they're also high-level—for single relations create, update, delete are both low and high level semantics. Expanding that to composing relations doesn't make sense, as all that does is extend the low-level semantics when those were just a coincidence in the first place.

**Comment by robinetmiller** [comment]
Everyone who read that in full is all the richer for it, so thanks for delving deep. :) For completeness, there's also the distillation @endash gave in Gitter that Repository can either be viewed as:
1. something intended to help us deal with relations
2. a design pattern that helps us isolate our persistence from our app  

I am very resoundingly in the mindset that 2 is the dominant benefit of Repo. This stems from:
- @solnic stating that he wants Repo to be the primary class used (at least for reads)
- General agreement that commands have _some_ sort of place within Repo (hence that this PR's purpose)
- A desire to reduce the number of interactive parts for common operations (as a Facade, or similar). This allows me to temporarily handwave away the persistence details while thinking about the domain solution (and vice versa)

And finally, a quote from a couple months back: 

> repositories are not “per model” and not “per gateway” they are “per environment”, so that you can use whatever relations you have there and produce whatever models you want from those relations
> -- @solnic 

Which _sounds_ like the number of Repos for a small to medium app is one, and then expands from there dependent on what's needed. 

So the statement above has been rattling in the back of my head as a guide for what Repo is / should be. In addition to that, multiple folk have expressed that being able to just define a domain-centric method name is really nice (eg. `#create_project_and_roles`). I also enjoy this feature, and would love to support & encourage that from the beginning to maturity. 

The intent for my sketch above was to provide a way for a single Repository to be able to service an entire small project, and then for the developer to decide to shear off another Repo as required. This is the way that most classes are developed anyway - adding on until clear lines are distinguishable, and then refactoring. 

To me, making it easy to do one-repo-per-model breaks that expandability out of the gate. The number of repos expands to match the number of models. I'm still not sure where the more complex methods would live. 

Another tilt at this, including some pies in some skies: 

``` ruby
repo.create_user(name: 'Jane')      # simple method "generated" by the API
repo.create_task(name: 'Laundry')   # same. It could pick apart the method name to know what relation to use. This enforces a standard by making it the easy path.

# restricted methods need a restriction to know which to change. 
# Maybe that's just the first param, and then there's a way to customize that? Not entirely satisfied here.
repo.update_user(1, name: 'Jane Doe')

# use the plural inflection to notice that we're doing a group. 
repo.update_users([2,3], [{name: 'John Doe'}, {name: 'Kermit the Frog'}])

# a more complex method, and would be defined by hand in Repo, 
# but it still looks familiar to the ones above - consistent is good
repo.update_users_and_tasks( nested_data_hash ) 
```

My ideal would be `repo.create(my_domain_object)` and have it somehow handle all the nesting and whatnot, but that might be a little _too_ abstract.

**Comment by kwando** [comment]
I just recently started to use `ROM::Repository`  so I might not have explored and discovered how I want to use it fully yet.

When I think about a _repository_ I'm think about it in the design pattern sense, an object that shields my application and domain logic from "database stuff". 

I consider `rom-repository` as a "database-stuff" building block which makes it a lot easier to implement your own application entity stores. 

In short I tend to view rom-repository as "something intended to help us deal with relations".

**Comment by endash** [comment]
Personally, I wouldn't subscribe to any hard and fast rules on number of repositories or what correspondence they should have to relations. A given repository would fill a particular role from the app's perspective of providing persistence in particular contexts with particular app-level domain constructs. Now, most of the time the number of repositories per relation will be conserved, but that isn't necessarily the case. The User/Admin dichotomy is a classic situation where you'd likely want two repos for a single relation. The relation is databasey implementation detail, the fact that both domain concepts, important enough to treat as distinct (IMO, any situation where you have different levels of privilege calls for separate classes. A surfeit of if tests, or relying on client code to only call some methods and not others, is asking for trouble), are ultimately persisted in a single table is immaterial.

It's important too that we're still defining classes, here... As in something that should have a single, well-defined responsibility.  I would not, for instance, pile methods onto a repository such that one client part of the app might use methods a, b, and c, while another client part might use methods d, e, and f. Classes, in my view, shouldn't be mere grab bags of disjointed methods, collected for convenience. Other SOLID principles are implicated as well... Dependency Inversion tells us that the interfaces our apps depend on should be owned by the app, with app-level semantics, rather than semantics dictated by low-level concerns of the dependency. Interface Segregation, as well—having one repository per app rather than logical domain concept would mean exposing a lot of extra interface almost everywhere it is used.

From my own perspective, I tend to relate a repository to the concept of an aggregate root. I'd start with the assumption of one repo per aggregate root, in which case a set of repositories, one per relation, is perfectly fine... If they're all valid aggregate roots. This stops being correct once you do start wanting to group relations into aggregates, but until that point it is perfectly alright, architecturally, to treat them as completely independent and orthogonal entities.

**Comment by endash** [comment]
@kwando I see the logic there but it strikes me as unnecessary regress... Low-level functionality that underlies our own implementations of the repository pattern is the sort of things that should be in core. An add-on library as a requirement merely for the ability to use a common design pattern is over complicating the project. Repositories that we implement DO need use of powerful, easy to use APIs for dealing with composed relations and whatnot... But that isn't part of the interface of a repository. Keep in mind that the model/crud repo class is "code free" so we're merely concerned with the public API from the app-side of things, not implementation (at present).

I do want it to be simple to write my hypothetical `create_admin_for_projects` repository method, composing relations and commands and whatnot. I don't want the repository to merely be another surface exposing relations and commands.

**Comment by solnic** [comment]
The reason why a single repo can access multiple relations is because you can construct a single aggregate from multiple relations _but_ a single repo should deal with the same _kind_ of aggregates. That's why I don't like the idea of having a single repo that is used as a way to deal with different _kinds_ of flat-objects and aggregates for both reading and writing. My idea behind `Repository::Model` was to make a clear distinction between a simple CRUD repo for a single _concept_ of your app which happens to match your database table (in sql case) and a more complicated one which can compose data from multiple relations.

Now, the tricky part, as always, are associations. Command graphs can handle nested input and they _might_ match the structures of relation graphs. Nested structures that your app receives rarely match the structure that you pull out of the db to return it to the client (ie in a web ui, or a json response). That's why I've become a fan of separating this logic _completely_.

Having said that, we should keep in mind that in common and simple cases an AR-like 1:1 match between structures stored in the database and structures exposed by your app are the same. This means you can have `a user with an address` and then have `users` relation and `addresses` relation with `user_id` FK. And you simply compose the two relations for reading and you also can create a user with an address using an input which is a nested hash with `{ user: { name: 'jane', address: { .. } }` structure. For cases like that, we can provide convenient APIs. We can have full CRUD for that including associations. Question is: where? `Repository`? Or should we create specialized repo subclasses for specific use-cases?

Here's an example:

``` ruby
class UserRepo < ROM::Repository
  relations :users, :addressses

  def [](id)
    user_with_address.by_id(id)
  end

  def create(user_data)
    command(:create, users).call(user_data)
  end

  def create_with_address(user_data)
    command(:create, user_with_address).call(user_data)
  end

  private

  def user_with_address
    users.combine_children(one: addresses)
  end
end

user_repo = UserRepo.new(rom)

user_repo.create(name: 'Jane')

user_repo.create_with_address(name: 'Jane', address: { city: 'Krakow' })

user_repo[1]
```

**Comment by solnic** [comment]
One more thing - what @endash wrote about underlying core interfaces is very true. Your app should not talk to ROM's interfaces at all, that's a general design principle we have, applies to everything. Things should be simple to compose, like I showed in the example above, but your repo exposes _your_ interfaces to the app. ie `Repository#command` would be private.

**Comment by endash** [comment]
To the extent that we're making assumptions of a 1:1 correspondence between the app domain and db schema, a la AR, that expectation should be pushed into a subclass. I don't see anything wrong in principle with being able to specify, in effect, a "user has many addresses" repo and get "create" and "create_with_address" for free. 

What's important from my perspective is:
1.  a repository has one root thing it works with, and all actions are with respect to that thing. IE, a users repository has "create," a generic repository doesn't have "create_user"
2.  simply combining more than one relation isn't a free for all on each relation. Our "users has many addresses" repository would NOT have an "update_address" method. There might possibly be a separate addresses repo for parts of code that work exclusively with addresses, otherwise modifying addresses is always part of modifying the user.
3.  the actions aren't too clever and don't do too much, as in taking a whole nested hash via the basic "create" and doing a bunch of creating of other relations... user still has to specify the semantics
4.  there's a fine line to walk between providing a bridge from AR or a basic auto-repository, and encouraging devs to view repositories as just a straight-to-DB access layer.

I would steer clear of using AR-ish association syntax/terminology for the API. One thought is to specify the desired composite commands as part of a more explicit, simple DSL, and then interpret the provided name as if it were an association. Something like

``` ruby
class UserRepo <  ROM::Repository::Auto
  relations :users, :addresses
  create :with_address
end
```

The first relation would be the root by default and would get the basic CRUD actions, and composite actions are specified explicitly. It provides a good mental model of "these are actual methods you would normally write, but because we're making some assumptions we'll handle that for you" more directly than writing "has_many :addresses" and suddenly getting all these methods by magic. There's a clear map from the magic to what it would look like if you had to implement it all directly.

**Comment by endash** [comment]
As far as the Model repo specifically, the major actual outstanding issue, from what I can tell, is what assumptions and syntax get us from

``` ruby
class User < ActiveRecord::Base
    has_many :addresses
    accepts_nested_attributes_for :addresses
end

class Address < ActiveRecord::Base
    belongs_to :user
end
```

to our `UserRepo`, and be able to do 

``` ruby
user_repo.create_with_addresses(user_hash)
```

with a minimum of pain (either short-term or long-term, via messed up mental models). @solnic in the gitter wrote this as an example of his current (private) API for repository commands:

``` ruby
create_user_with_task = repo.command(:create,  repo.users.combine_children(one: repo.tasks))
```

I propose that (base) Repository become opinionated when it comes to how relations can be composed. Rather than a basic `relations` array that just gives us access to each relation at a low level, the repo should have an explicit root relation, and separately any associated relations. Repository could then restrict access to the non-root relation to prevent frankenstein repos. Then, we could specify the commands on (model) Repository that we want auto-generated as a map from the root relation:

``` ruby
class UserRepo < Repository::Model
    relation :users => :addresses
    create :with_addresses, :many => :addresses

    # basic create, update, etc for users would not have to be specified
end
```

Maybe something more complex would be (bad example, but I can't immediately think of a good 3-tier aggregate)

``` ruby
class UserRepo < Repository::Model
    relation :users => [:orders, :order_items]
    create :with_order, :many => {:orders => {:many => :order_items}}
end
```

Then it's a simple matter of parsing the provided map for each command into `combine_children` calls. For anything more complex than single-relation stuff the user provides the semantics. I don't think this is optimal currently because I can see with just a second command (for update), replicating the association map from line to line will get heavy.

**Comment by solnic** [comment]
I like those DSL suggestions although I must say I'll be interested in finishing the `Repository#command` interface but way less interested in working on class-level macros. I wouldn't be against them, though.

**Comment by endash** [comment]
@solnic Once the underpinnings are in place I'll take a look at implementing them. The current improvements are great themselves but fall short of the turnkey simplicity needed to pitch it as an easy migration for AR devs. Which actually I've over-focused on, and isn't the be-all-and-end-all of considerations, but still no reason to stop just short of the finish line on that goal.

**Comment by solnic** [comment]
That would be awesome Chris. Thank you.

**Comment by solnic** [comment]
Closing in favor of #15
